File: panopticon-event-logger/src/main/java/com/example/panopticon_event_logger/model/EventLoggerModel.java
package com.example.panopticon_event_logger.model;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import org.jetbrains.annotations.NotNull;

import java.util.HashMap;

public class EventLoggerModel {

    @SerializedName("ip")
    @Expose
    private String ip;
    @SerializedName("app_id")
    @Expose
    private String appId ;
    @SerializedName("app_name")
    @Expose
    private String appName;
    @SerializedName("event_name")
    @Expose
    private String eventName;
    @SerializedName("user_id")
    @Expose
    private int userId ;
    @SerializedName("user_type")
    @Expose
    private String userType ;
    @SerializedName("device_model")
    @Expose
    private String deviceModel;
    @SerializedName("device_id")
    @Expose
    private String deviceId ;
    @SerializedName("log_time")
    @Expose
    private String logTime ;
    @SerializedName("event_data")
    @Expose
    private HashMap<String, String> eventData ;
    @SerializedName("app_version")
    @Expose
    private String appVersion;

    public String getIp() {
        return ip;
    }

    public String getAppId() {
        return appId;
    }

    public String getAppName() {
        return appName;
    }

    public String getEventName() {
        return eventName;
    }

    public int getUserId() {
        return userId;
    }

    public String getUserType() {
        return userType;
    }

    public String getDeviceModel() {
        return deviceModel;
    }

    public String getDeviceId() {
        return deviceId;
    }

    public String getLogTime() {
        return logTime;
    }

    public HashMap<String, String> getEventData() {
        return eventData;
    }

    public String getAppVersion() {
        return appVersion;
    }

    private EventLoggerModel(EventLoggerBuilder builder) {
        this.ip=builder.ip;
        this.appId=builder.appId;
        this.appName=builder.appName;
        this.eventName=builder.eventName;
        this.userId=builder.userId;
        this.userType=builder.userType;
        this.deviceModel=builder.deviceModel;
        this.deviceId=builder.deviceId;
        this.logTime=builder.logTime;
        this.eventData=builder.eventData;
        this.appVersion=builder.appVersion;
    }

    //Builder Class

    public static class EventLoggerBuilder{

        public EventLoggerBuilder() {
        }

        private String ip;
        private String appId ;
        private String appName;
        private String eventName;
        private int userId ;
        private String userType ;
        private String deviceModel ;
        private String deviceId ;
        private String logTime ;
        private HashMap<String, String> eventData = new HashMap<String, String>();
        private String appVersion ;


        public EventLoggerBuilder setIp(String ip) {
            this.ip = ip;
            return this ;
        }

        public EventLoggerBuilder setAppId(String appId) {
            this.appId = appId;
            return this ;
        }

        public EventLoggerBuilder setAppName(String appName) {
            this.appName = appName;
            return this ;
        }

        public EventLoggerBuilder setEventName(String eventName) {
            this.eventName = eventName;
            return this ;
        }

        public EventLoggerBuilder setUserId(int userId) {
            this.userId = userId;
            return this ;
        }

        public EventLoggerBuilder setUserType(String userType) {
            this.userType = userType;
            return this ;
        }

        public EventLoggerBuilder setDeviceModel(String deviceModel) {
            this.deviceModel = deviceModel;
            return this ;
        }

        public EventLoggerBuilder setDeviceId(String deviceId) {
            this.deviceId = deviceId;
            return this ;
        }

        public EventLoggerBuilder setLogTime(String logTime) {
            this.logTime = logTime;
            return this ;
        }

        public EventLoggerBuilder setEventData(HashMap<String, String> eventData) {
            this.eventData = eventData;
            return this ;
        }

        public EventLoggerBuilder setAppVersion(String appVersion) {
            this.appVersion = appVersion;
            return this ;
        }

        public EventLoggerModel build(){
            return new EventLoggerModel(this);
        }
    }


    @NotNull
    @Override
    public String toString() {
        return "EventLoggerModel{" +
                "ip='" + ip + '\'' +
                ", appId='" + appId + '\'' +
                ", appName='" + appName + '\'' +
                ", eventName='" + eventName + '\'' +
                ", userId=" + userId +
                ", userType='" + userType + '\'' +
                ", deviceModel='" + deviceModel + '\'' +
                ", deviceId='" + deviceId + '\'' +
                ", logTime='" + logTime + '\'' +
                ", eventData=" + eventData +
                ", appVersion='" + appVersion + '\'' +
                '}';
    }
}

File: panopticon-event-logger/src/main/java/com/example/panopticon_event_logger/model/EventLoggerModel.kt
package com.example.panopticon_event_logger.model

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import java.util.*

class EventLoggerModel private constructor(builder: EventLoggerBuilder) {
    @SerializedName("ip")
    @Expose
    val ip: String?
    @SerializedName("app_id")
    @Expose
    val appId: String?
    @SerializedName("app_name")
    @Expose
    val appName: String?
    @SerializedName("event_name")
    @Expose
    val eventName: String?
    @SerializedName("user_id")
    @Expose
    val userId: Int
    @SerializedName("user_type")
    @Expose
    val userType: String?
    @SerializedName("device_model")
    @Expose
    val deviceModel: String?
    @SerializedName("device_id")
    @Expose
    val deviceId: String?
    @SerializedName("log_time")
    @Expose
    val logTime: String?
    @SerializedName("event_data")
    @Expose
    val eventData: HashMap<String, String>
    @SerializedName("app_version")
    @Expose
    val appVersion: String?

    //Builder Class
    class EventLoggerBuilder {
        var ip: String? = null
        var appId: String? = null
        var appName: String? = null
        var eventName: String? = null
        var userId = 0
        var userType: String? = null
        var deviceModel: String? = null
        var deviceId: String? = null
        var logTime: String? = null
        var eventData = HashMap<String, String>()
        var appVersion: String? = null

        fun setIp(ip: String?): EventLoggerBuilder {
            this.ip = ip
            return this
        }

        fun setAppId(appId: String?): EventLoggerBuilder {
            this.appId = appId
            return this
        }

        fun setAppName(appName: String?): EventLoggerBuilder {
            this.appName = appName
            return this
        }

        fun setEventName(eventName: String?): EventLoggerBuilder {
            this.eventName = eventName
            return this
        }

        fun setUserId(userId: Int): EventLoggerBuilder {
            this.userId = userId
            return this
        }

        fun setUserType(userType: String?): EventLoggerBuilder {
            this.userType = userType
            return this
        }

        fun setDeviceModel(deviceModel: String?): EventLoggerBuilder {
            this.deviceModel = deviceModel
            return this
        }

        fun setDeviceId(deviceId: String?): EventLoggerBuilder {
            this.deviceId = deviceId
            return this
        }

        fun setLogTime(logTime: String?): EventLoggerBuilder {
            this.logTime = logTime
            return this
        }

        fun setEventData(eventData: HashMap<String, String>): EventLoggerBuilder {
            this.eventData = eventData
            return this
        }

        fun setAppVersion(appVersion: String?): EventLoggerBuilder {
            this.appVersion = appVersion
            return this
        }

        fun build(): EventLoggerModel {
            return EventLoggerModel(this)
        }
    }

    override fun toString(): String {
        return "EventLoggerModel{" +
                "ip='" + ip + '\'' +
                ", appId='" + appId + '\'' +
                ", appName='" + appName + '\'' +
                ", eventName='" + eventName + '\'' +
                ", userId=" + userId +
                ", userType='" + userType + '\'' +
                ", deviceModel='" + deviceModel + '\'' +
                ", deviceId='" + deviceId + '\'' +
                ", logTime='" + logTime + '\'' +
                ", eventData=" + eventData +
                ", appVersion='" + appVersion + '\'' +
                '}'
    }

    init {
        ip = builder.ip
        appId = builder.appId
        appName = builder.appName
        eventName = builder.eventName
        userId = builder.userId
        userType = builder.userType
        deviceModel = builder.deviceModel
        deviceId = builder.deviceId
        logTime = builder.logTime
        eventData = builder.eventData
        appVersion = builder.appVersion
    }
}